### PR TITLE
PorousFlow: utils for hysteretic relative permeabilities added

### LINKED ISSUE #16296: PorousFlow: create hysteretic relative permeability utils
## Reason

To perform simulations with hysteretic relative permeability curves

## Design

This issue will add utility functions to porous_flow/src/utils that will provide the liquid and gas relative permeability values, and their first and second derivatives, given the liquid saturation, van Genuchten parameters, and extension parameters.

## Impact

These additions will not be available to modellers.  Next week's addition will hopefully provide them with access to the hysteretic rel perms.

### PR DESCRIPTION
**Pull request by WilkAndy**
Also, a few minor errors corrected in hys.py

Fixes #16296

### CODE REVIEW

**Comment by WilkAndy** [comment]
Hey @cpgr.  Here's the next installment in the hysteretic stuff - utils for computing the relative permeabilities.  Could you review at some stage, please?

**Comment by cpgr** [comment]
Will take a look tonight

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C [hunk 1]
```diff
@@ -609,4 +609,268 @@ d2saturationHys(Real pc,
   }
   return d2s;
 }
+
+Real
+cubic(Real x, Real x0, Real y0, Real y0p, Real x1, Real y1, Real y1p)
+{
```
Comment (cpgr) [inline modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:615]: Should you check that `x1 > x0`, maybe an assert?

Comment (cpgr) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:615]: Certainly going to fail if `x1=x0`

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C [hunk 2]
```diff
@@ -609,4 +609,268 @@ d2saturationHys(Real pc,
...
+                        Real sgrmax,
+                        Real sldel,
+                        Real m,
+                        Real upper_liquid_param,
+                        Real y0,
+                        Real y0p,
+                        Real y1,
+                        Real y1p)
+{
+  if (sl < slr) // by the definition of slr, always return 0
```
Comment (cpgr) [inline modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:664]: Could this be `sl<= sir`? I think if `sl==slr` that `kr=0` too?

Comment (cpgr) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:664]: oops - auto corrected `slr` to `sir`!

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C [hunk 3]
```diff
@@ -609,4 +609,268 @@ d2saturationHys(Real pc,
...
+                         Real sgrmax,
+                         Real sldel,
+                         Real m,
+                         Real upper_liquid_param,
+                         Real y0,
+                         Real y0p,
+                         Real y1,
+                         Real y1p)
+{
+  if (sl < slr) // by the definition of slr, always return 0
```
Comment (cpgr) [inline modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:720]: Same comment as above

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C [hunk 4]
```diff
@@ -609,4 +609,268 @@ d2saturationHys(Real pc,
...
+      const Real c_prime = c / m / (sl_bar + s_gt_bar) * (sl_bar_prime + s_gt_bar_prime);
+      a = (1 - s_gt_bar / (1.0 - sl_bar_del)) * std::pow(1.0 - c, m);
+      a_prime =
+          -s_gt_bar_prime / (1.0 - sl_bar_del) * std::pow(1.0 - c, m) - m * a / (1.0 - c) * c_prime;
+      b = s_gt_bar / (1.0 - sl_bar_del) * std::pow(1.0 - std::pow(sl_bar_del, 1.0 / m), m);
+      b_prime = s_gt_bar_prime * b / s_gt_bar;
+    }
+  }
+  const Real kr = std::sqrt(sl_bar) * Utility::pow<2>(1.0 - a - b);
+  return 0.5 * kr / sl_bar * sl_bar_prime -
```
Comment (cpgr) [inline modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:779]: Early return for `sl = slr` would avoid this blowing up when `sl_bar=0`.

**Review by cpgr — COMMENTED** [review]
I didn't really read the unit tests but I assume that they test all cases. 

My only other comment is that the cubic extension function might eventually be useful elsewhere so one day it might get moved to some other util file.

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C (hunk 1 above)
Comment (WilkAndy) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:615]: Included an assert, and made a new `PorousFlowCubic` namespace as suggested

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C (hunk 2 above)
Comment (WilkAndy) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:664]: Yea `<=` is better, i agree

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C (hunk 3 above)
Comment (WilkAndy) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:720]: agree

File: modules/porous_flow/src/utils/PorousFlowVanGenuchten.C (hunk 4 above)
Comment (WilkAndy) [reply modules/porous_flow/src/utils/PorousFlowVanGenuchten.C:779]: oooh, that's mean.  Yea, i agree.

**Comment by WilkAndy** [comment]
I think this is ready for merge, @aeslaughter .

**Comment by WilkAndy** [comment]
Hey @loganharbour , just wondering if you're still awake and can merge this?  I'm currently extending this capability ( #16324 ) and don't want to push up to this PR to make it too huge.

**Review by loganharbour — APPROVED** [review]

**Comment by WilkAndy** [comment]
Thanks for the merge, @loganharbour ,

a

**Pull request merged**